chips/sifive/src/prci.rs

```diff
@@ -48,29 +51,111 @@ register_bitfields![u32,
 
 pub enum ClockFrequency {
     Freq16Mhz,
+    Freq320Mhz,
 }
 
 pub struct Prci {
     registers: StaticRef<PrciRegisters>,
+    current_frequency: Cell<ClockFrequency>,
 }
 
 impl Prci {
     pub const fn new(base: StaticRef<PrciRegisters>) -> Prci {
-        Prci { registers: base }
+        Prci {
+            registers: base,
+            current_frequency: Cell::new(ClockFrequency::Freq16Mhz),
+        }
+    }
+
+    pub fn switch_to_internal_clock(&self) {
+        let regs = self.registers;
+        // Enable internal high-frequency clock if it's not enabled
+        if regs.hfrosccfg.read(hfrosccfg::enable) == 0 {
+            regs.hfrosccfg.modify(hfrosccfg::enable::SET);
+        }
+        // ... Wait until the clock is ready
+        while regs.hfrosccfg.read(hfrosccfg::ready) == 0 {}
+        // ... and now actually switch
+        regs.pllcfg
+            .modify(pllcfg::sel::CLEAR + pllcfg::bypass::CLEAR);
+    }
+
+    pub fn set_internal_clock_default(&self) {
+        let regs = self.registers;
+        // Set to defaults, which according to data sheet should set to 14.4MHz +- 50%,
+        regs.hfrosccfg
+            .modify(hfrosccfg::div.val(4) + hfrosccfg::trim.val(0x10));
+    }
+
+    pub fn enable_external_clock(&self) {
+        let regs = self.registers;
+        // Make sure external crystal oscillator is enabled
+        if regs.hfxosccfg.read(hfxosccfg::enable) == 0 {
+            regs.hfxosccfg.modify(hfxosccfg::enable::SET);
+        }
+        // ... Wait until the clock is ready
+        while regs.hfxosccfg.read(hfxosccfg::ready) == 0 {}
     }
 
     pub fn set_clock_frequency(&self, frequency: ClockFrequency) {
         let regs = self.registers;
+        // According to someone affiliated with SiFive in forum post:
+        // https://forums.sifive.com/t/is-it-possible-to-brick-the-hifive-board/751/6,
+        // it is safe to adjust the internal high frequency clock while it
+        // is in use, but not the PLL output.
+        //
+        // So first switch to internal clock before doing any other clock manipulation
+
+        // Reset internal clock to defaults so we can estimate how long we're spinning for the PLL
+        // lock delay. At default, the frequency should be 14.4MHz +- 50%, so a maximum frequency
+        // of just under 22MHz
+        self.set_internal_clock_default();
+        self.switch_to_internal_clock();
+
+        // Make sure external clock subsystem is enabled before adjusting the PLL
+        self.enable_external_clock();
 
         match frequency {
             ClockFrequency::Freq16Mhz => {
+                // Bypass enabled, feeds clock directly from external clock. For HiFive1 revB, this
+                // is a 16 MHz clock
                 regs.pllcfg
                     .modify(pllcfg::bypass::SET + pllcfg::refsel::SET);
```

Comment:
One thing that bothers me here is the implicit assumption that all SiFive chips will be running off a 16MHz external clock. Should we, instead, provide some kind of input parameter when creating the peripheral indicating the value of the external clock as installed on the board? This way different SiFive chips on multiple boards can specify what their actual external clocks are.

Reply:
You can have an input parameter, we do that in [OpenTitan](https://github.com/tock/tock/blob/master/chips/earlgrey/src/chip_config.rs) for the UART speeds for example.

If you want to feel free to open a PR. The problem with the SiFive SoCs is that they aren't [great Tock candidates](https://github.com/tock/libtock-rs/issues/156) so there hasn't been a lot of interest.

Reply:
For now, per another comment by @bradjc , I think we can assume that 16MHz external clock. If in the future we support another SiFive chip based board with a different external clock, we can add that later, I think.